tsdb/memdb/database.go

```diff
@@ -0,0 +1,115 @@
+package memdb
+
+import (
+	"context"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+const (
+	// evictor evicts the stores in this interval
+	evictorInterval = time.Second * 30
+	// store will be evicted if have not been used in such a duration, unit: second
+	evictingDuration = 300
+	// concurrency of ts-store-evictor
+	evictorsOfTsStore = 5
+)
+
+// MemoryDatabase is a database-like concept of Shard as memTable in cassandra.
+type MemoryDatabase struct {
+	// key: measurement-name
+	measurementsMap map[string]*MeasurementStore
+	// RWMutex for accessing measurementsMap.
+	mu sync.RWMutex
+}
+
+// NewMemDatabase returns a new MemoryDatabase.
+func NewMemDatabase(ctx context.Context) *MemoryDatabase {
+	md := MemoryDatabase{
+		measurementsMap: make(map[string]*MeasurementStore)}
+	go md.evictorRunner(ctx)
+	return &md
+}
+
+// GetTimeSeriesStore returns a TimeSeriesStore by measurement + tags.
+func (md *MemoryDatabase) GetTimeSeriesStore(measurement string, tags string) *TimeSeriesStore {
+	md.mu.RLock()
+	store, exist := md.measurementsMap[measurement]
+	md.mu.RUnlock()
+
+	if !exist {
+		md.mu.Lock()
+		store, exist = md.measurementsMap[measurement]
+		if !exist {
+			store = newMeasurement(measurement)
+			md.measurementsMap[measurement] = store
+		}
+		md.mu.Unlock()
+	}
+	return store.getTimeSeries(tags)
+}
+
+// evictorRunner runs evictor periodically.
+func (md *MemoryDatabase) evictorRunner(ctx context.Context) {
+	// delay starting evictor in random 0-30 seconds.
+	delayRandom := time.NewTimer(time.Millisecond * time.Duration(
+		rand.New(rand.NewSource(time.Now().UnixNano())).Intn(3000*10)))
+	select {
+	case <-ctx.Done():
+		delayRandom.Stop()
+		return
+	case <-delayRandom.C:
+		delayRandom.Stop()
+	}
+
+	evictorTicker := time.NewTicker(evictorInterval)
+	defer evictorTicker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-evictorTicker.C:
+			mStoreCh := make(chan *MeasurementStore, evictorsOfTsStore)
+			go md.mStoreProducer(ctx, mStoreCh)
+			for i := 0; i < evictorsOfTsStore; i++ {
+				go md.mStoreEvictor(mStoreCh)
+			}
+		}
+	}
+}
+
+// mStoreProducer gets all measurement-stores, and send them to channel for mStoreEvictor to check.
+func (md *MemoryDatabase) mStoreProducer(ctx context.Context, mStoreCh chan<- *MeasurementStore) {
+	// get all md
+	md.mu.RLock()
+	allMstore := make([]*MeasurementStore, len(md.measurementsMap))
+	offset := 0
+	for _, mStore := range md.measurementsMap {
+		allMstore[offset] = mStore
+		offset++
+	}
+	md.mu.RUnlock()
+
+	defer close(mStoreCh)
+	for _, mStore := range allMstore {
+		select {
+		case <-ctx.Done():
+			return
+		case mStoreCh <- mStore:
+		}
+	}
+}
+
+// mStoreEvictor evict tsStore of mStore concurrently,
+// and delete MeasurementStore whose timeSeriesMap is empty.
+func (md *MemoryDatabase) mStoreEvictor(mStoreCh <-chan *MeasurementStore) {
+	for mStore := range mStoreCh {
+		isEmtpy := mStore.evict()
+		if isEmtpy {
+			md.mu.Lock()
```

Comment:
1. not threadsafe;
2.mStore.evict() return true, other goroutine write data, after write do delete ops